x-pack/solutions/observability/plugins/streams/server/lib/streams/client.ts

```diff
@@ -739,38 +791,124 @@ export class StreamsClient {
+  /**
+   * Returns the unwired root stream for a wired stream.
+   *
+   * For a stream wired to logs, this returns undefined.
+   * For a stream wired to an unwired classic stream, this returns the classic stream.
+   *
+   * This is done by fetching all classic stream definitions that start with the first dot part of the dataset
+   * and looking for the longest match (in terms of numbers of dots in the dataset).
+   */
+  async getUnwiredRootForStream(
+    definition: WiredStreamDefinition
+  ): Promise<StreamDefinition | undefined> {
+    let classicRoot: StreamDefinition | undefined;
+    let classicRootSegmentLength = 0;
+    const streamName = parseStreamName(definition.name);
+    if (streamName.type === 'dsns') {
+      const streams = await this.getStreamDefinitions({
+        query: {
+          bool: {
+            filter: [
+              {
+                prefix: {
+                  name: `${streamName.datastreamType}-${streamName.datasetSegments[0]}`,
+                },
+              },
+            ],
+          },
+        },
+      });
+
+      streams
+        .filter((candidateStream) => !isWiredStream(candidateStream))
+        .forEach((candidateStream) => {
+          const candidateStreamName = parseStreamName(candidateStream.name);
+          // The longest unwired stream with a matching prefix is the classic root
+          if (
+            candidateStreamName.type === 'dsns' &&
+            candidateStreamName.datasetSegments.every(
+              (segment, index) => segment === streamName.datasetSegments[index]
+            ) &&
+            candidateStreamName.datastreamType === streamName.datastreamType &&
+            candidateStreamName.datastreamNamespace === streamName.datastreamNamespace &&
+            candidateStreamName.datasetSegments.length > classicRootSegmentLength
+          ) {
+            classicRoot = candidateStream;
+            classicRootSegmentLength = candidateStreamName.datasetSegments.length;
+          }
+        });
+    }
+    return classicRoot;
+  }
+
+  async getAncestors(definition: StreamDefinition): Promise<StreamDefinition[]> {
+    let unwiredRoot = isWiredStream(definition)
+      ? await this.getUnwiredRootForStream(definition)
+      : undefined;
 
-    return this.getManagedStreams({
+    const ancestorIds = getAncestors(definition.name, unwiredRoot?.name);
+
+    const ancestors = await this.getStreamDefinitions({
       query: {
         bool: {
           filter: [{ terms: { name: ancestorIds } }],
         },
       },
-    }).then((streams) => streams.filter(isWiredStream));
+    });
+    const streamName = parseStreamName(definition.name);
+    if (isWiredStream(definition) && streamName.type === 'dsns' && ancestors.length === 0) {
+      // We have a DSNS stream, which means it is not connected to the root logs stream, but no ancestors.
+      // This can happen if the classic stream for the unwired root hasn't been created yet.
```

Comment:
```suggestion
      // This can happen if the classic stream definition for the unwired root hasn't been created yet.
```

Right? Because the (data) stream exists.

Reply:
right, should have been more exact